test/app-luatest/gh_9806_http_client_version_test.lua

```diff
@@ -0,0 +1,95 @@
+local t = require('luatest')
+local treegen = require('test.treegen')
+local justrun = require('test.justrun')
+local socket = require('socket')
+
+local g = t.group()
+
+g.before_all(treegen.init)
+g.after_all(treegen.clean)
+
+g.before_each(function(g)
+    g.server = socket.tcp_server('127.0.0.1', 0, function(_s) end)
+end)
+
+g.after_each(function(g)
+    g.server:close()
+end)
+
+local http_request_script = string.dump(function()
+    local http_client = require('http.client').new()
+    local proto = os.getenv('PROTO')
+    local port = os.getenv('HTTP_SERVER_PORT')
+    local uri = proto .. '://127.0.0.1:' .. port
+    pcall(http_client.get, http_client, uri, {
+        verbose = true,
+        http_version = os.getenv('HTTP_VERSION'),
+    })
+end)
+
+g.test_http_version = function()
+    local dir = treegen.prepare_directory(g, {}, {})
+    local script_name = 'http_request.lua'
+    treegen.write_script(dir, script_name, http_request_script)
```

Comment:
To be honestly, I don't get why do you need `treegen` here.
After removing `treegen` env variables will be gone too:

```lua
 local http_req = string.dump(function(port, version)
    local http_client = require('http.client').new()
    local uri = 'http://127.0.0.1:' .. port
    pcall(http_client.get, http_client, uri, {
        verbose = true,
        http_version = version,
    })
end)

assert(loadstring(http_req))('35249', '1.0')
```

Reply:
The main problem is to catch `stderr`.

We can perform the usual `dup/dup2/pipe` catching from the same process. However, luatest performs something like this on its own and I don't want to clash here. In theory, the double wrapping should work, but it seems dangerous for me.

So, we need a separate process. It means some `popen` (we use `justrun` wrapper here). That means that we need to run `tarantool some_script.lua`.

We can place this script into the source code or generate in runtime. I prefer the second way, because it allows to keep all the code that relates to a test case in the same file.

Since we generate the file in runtime, we need a temporary directory to store it. We can create it using `fio.tempdir()` and remove using `fio.rmtree()` in an `after_all` hook to don't leave garbage after each run of the testing.

In fact, it is what `treegen` is doing. It creates and tracks temporary directories and provides some helpers for creating scripts in runtime.